packages/gatsby/src/utils/sanitize-node.ts

```diff
@@ -20,15 +68,15 @@ const sanitizeNode = (data, isNode = true, path = new Set()) => {
         returnData[key] = o
         return
       }
-      returnData[key] = sanitizeNode(o, false, path)
+      returnData[key] = sanitizeNode(o as data, false, path)
 
       if (returnData[key] !== o) {
         anyFieldChanged = true
       }
     })
 
     if (anyFieldChanged) {
-      data = omitUndefined(returnData)
+      data = omitUndefined(returnData as data) as data
```

Comment:
If we can type `returnData` during initialization, we should be able to get away with no casting here.

Reply:
`returnData` is initalised as `{}` or `[]`.
- `{}` is not a valid `IGatsbyNode` error: Type '{}' is missing the following properties from type 'IGatsbyNode': id, parent, children, internal, fields
- `[]` is not a valid `GatsbyIterable<IGatsbyNode>` error: 
Type 'never[]' is missing the following properties from type 'GatsbyIterable<IGatsbyNode>': source, deduplicate, mergeSorted, intersectSorted, deduplicateSorted

I could cast when we define `returnData` to avoid the double casting here, but we would still need one as `omitUndefined` returns a `Partial<data>` (from lodash, pickBy returns a `Partial`).

I think it's down to personal preference here, I feel the double cast here is more significant than having it when we define `returnData` but I'm open to change it.

Reply:
Sounds reasonable enough. Thank you for the detailed response. 😄